library-best-practices/JLBP-8.md

```diff
@@ -0,0 +1,11 @@
+[JLBP-8] Advance widely used functionality to a stable version
+--------------------------------------------------------------
+
+Libraries at 0.x versions and unstable features marked with
+annotations make no promises about stability, and if they are widely
+used, they can cause a lot of problems when there are surface
+breakages (which is allowed per the rules of unstable functionality,
+explained in [JLBP-3](JLBP-3.md)). If such breakages can cause
+widespread pain, it can be seen as "common-law GA"; the preferable
+path is to release the library as-is, with all its warts, maybe
+marking less-desirable elements as deprecated.
```

Comment:
Releasing deprecated as 1.0 smells really bad. I don't think we need to mention the deprecation here.

Reply:
What I would like to recommend here is that the question in advance of 1.0 is, are we happy with what we have? Not "What do we still want to add?"

When a library has useful functionality and a stable API, it should be released. Do not wait for every feature to be finished.

Reply:
Sure, releasing deprecated at 1.0 smells bad, but it could be extremely disruptive to rip out something that's deprecated if it is widely used at multiple levels of a dependency tree; the value of having a stable artifact might be higher than the negative of deprecated features in the stable artifact.

I completely agree with the guidance that functionality doesn't have to be finished, it just has to be a frozen surface.